Mathlib/Analysis/NormedSpace/MultipliableUniformlyOn.lean

```diff
@@ -0,0 +1,151 @@
+
+lemma Summable.tendstoUniformlyOn_tsum_nat_log_one_add {f : ℕ → α → ℂ}
+    {u : ℕ → ℝ} (hu : Summable u) (h : ∀ᶠ n in atTop, ∀ x ∈ K, ‖f n x‖ ≤ u n) :
+    TendstoUniformlyOn (fun n x ↦ ∑ i ∈ Finset.range n, log (1 + f i x))
+    (fun x ↦ ∑' i : ℕ, Complex.log (1 + f i x)) atTop K := by
+  rw [← Nat.cofinite_eq_atTop] at h
+  exact (hu.hasSumUniformlyOn_log_one_add h).tendstoUniformlyOn_finset_range rfl
+
+end summable
+
+section UniformlyOn
+
+/-Note this is false without hfn. -/
+lemma hasProdUniformlyOn_of_clog {f : ι → α → ℂ} {𝔖 : Set (Set α)}
+    (hf : SummableUniformlyOn (fun i x ↦ log (f i x)) 𝔖) (hfn : ∀ K ∈ 𝔖, ∀ x ∈ K, ∀ i, f i x ≠ 0)
+    (hg : ∀ K ∈ 𝔖, BddAbove <| (fun x ↦ (∑' n, log (f n x)).re) '' K) :
+    HasProdUniformlyOn f (fun x ↦ ∏' i, f i x) 𝔖 := by
+  simp only [hasProdUniformlyOn_iff_tendstoUniformlyOn, Set.mem_singleton_iff, forall_eq]
+  obtain ⟨r, hr⟩ := hf.exists
+  intro K hK
+  suffices H : TendstoUniformlyOn (fun s a ↦ ∏ i ∈ s, f i a) (cexp ∘ r) atTop K by
+    apply TendstoUniformlyOn.congr_right H
+    apply Set.EqOn.trans (Set.EqOn.comp_left (hr.tsum_eqOn hK)).symm
+    exact fun x hx ↦ (cexp_tsum_eq_tprod (hfn K hK x hx) (hf.summable hK hx))
+  have h1 := hr.tsum_eqOn hK
+  simp only [hasSumUniformlyOn_iff_tendstoUniformlyOn, Set.mem_singleton_iff, forall_eq] at hr
+  refine ((hr K hK).comp_cexp ?_).congr ?_
+  · simp +contextual [← h1 _]
+    exact hg K hK
+  · filter_upwards with s i hi using by simp [exp_sum, fun y ↦ exp_log (hfn K hK i hi y)]
+
+lemma multipliableUniformlyOn_of_clog {f : ι → α → ℂ} {𝔖 : Set (Set α)}
+    (hf : SummableUniformlyOn (fun i x ↦ log (f i x)) 𝔖) (hfn : ∀ K ∈ 𝔖, ∀ x ∈ K, ∀ i, f i x ≠ 0)
+    (hg : ∀ K ∈ 𝔖, BddAbove <| (fun x ↦ (∑' n, log (f n x)).re) '' K) :
+    MultipliableUniformlyOn f 𝔖 :=
+    ⟨_, hasProdUniformlyOn_of_clog hf hfn hg⟩
+
+namespace Summable
+
+variable {R : Type*} [NormedCommRing R] [NormOneClass R] [CompleteSpace R] [TopologicalSpace α]
+  {f : ι → α → R} {K : Set α} {u : ι → ℝ}
+
+lemma hasProdUniformlyOn_one_add (hK : IsCompact K) (hu : Summable u)
+    (h : ∀ᶠ n in cofinite, ∀ x ∈ K, ‖f n x‖ ≤ u n) (hcts : ∀ n, ContinuousOn (f n) K) :
+    HasProdUniformlyOn (fun n x ↦ 1 + f n x) (fun x ↦ ∏' i, (1 + f i x)) {K} := by
+  simp only [hasProdUniformlyOn_iff_tendstoUniformlyOn, Set.mem_singleton_iff, forall_eq,
+    tendstoUniformlyOn_iff_tendstoUniformly_comp_coe]
+  by_cases hKe : K = ∅
+  · simp [TendstoUniformly, hKe]
+  · haveI hCK : CompactSpace K := isCompact_iff_compactSpace.mp hK
+    haveI hne : Nonempty K := by rwa [Set.nonempty_coe_sort, Set.nonempty_iff_ne_empty]
+    let f' n : C(K, R) := ⟨_, continuousOn_iff_continuous_restrict.mp (hcts n)⟩
+    have hf'_bd : ∀ᶠ (n : ι) in cofinite, ‖f' n‖ ≤ u n := by
+      simp_rw [ContinuousMap.norm_le_of_nonempty]
+      filter_upwards [h] with n hn using fun x ↦ hn x x.2
+    have hM : Multipliable fun i ↦ 1 + f' i := by
+      apply _root_.multipliable_one_add_of_summable
```

Comment:
We should add the `protected` flag to `Complex.multipliable_one_add_of_summable` so we don't need `_root_` here.